lib/tlIO/FFmpegWrite.cpp

```diff
@@ -42,198 +176,582 @@ namespace tl
+                    &p.avAudioCodecContext->ch_layout, buf, 256);
+                    
+                if (p.sampleRate != info.audio.sampleRate || resample)
+                {
+                    const audio::Info& input = info.audio;
+                    audio::Info output(info.audio.channelCount,
+                                       toAudioType(p.avAudioCodecContext->sample_fmt),
+                                       p.sampleRate);
+                    p.resample = audio::AudioResample::create(input, output);
+                    
+                    if (auto logSystem = _logSystem.lock())
+                    {
+                        logSystem->print(
+                            "tl::io::ffmpeg::Plugin::Write",
+                            string::Format(
+                                "Resample from layout {0}, {1} channels, type "
+                                "{2}, sample rate {3} to layout {4}, {5} "
+                                "channels, type {6}, sample rate {7}.")
+                            .arg(buf)
+                            .arg(input.channelCount)
+                            .arg(input.dataType)
+                            .arg(input.sampleRate)
+                            .arg(buf)
+                            .arg(output.channelCount)
+                            .arg(output.dataType)
+                            .arg(output.sampleRate));
+                    }
+                }
+                else
+                {
+                    const audio::Info& input = info.audio;
+                    if (auto logSystem = _logSystem.lock())
+                    {
+                        logSystem->print(
+                            "tl::io::ffmpeg::Plugin::Write",
+                            string::Format(
+                                "Save from layout {0}, {1} channels, type "
+                                "{2}, sample rate {3}.")
+                            .arg(buf)
+                            .arg(input.channelCount)
+                            .arg(input.dataType)
+                            .arg(input.sampleRate));
+                    }
+                }
 
-            r = avio_open(&p.avFormatContext->pb, p.fileName.c_str(), AVIO_FLAG_WRITE);
-            if (r < 0)
-            {
-                throw std::runtime_error(string::Format("{0}: {1}").arg(p.fileName).arg(getErrorLabel(r)));
-            }
+                p.avAudioCodecContext->bit_rate = 69000;
+                p.avAudioCodecContext->sample_rate = p.sampleRate;
+                p.avAudioCodecContext->time_base.num = 1;
+                p.avAudioCodecContext->time_base.den = p.sampleRate;
 
-            r = avformat_write_header(p.avFormatContext, NULL);
-            if (r < 0)
-            {
-                throw std::runtime_error(string::Format("{0}: {1}").arg(p.fileName).arg(getErrorLabel(r)));
-            }
+                if ((p.avAudioCodecContext->block_align == 1 ||
```

Comment:
What are these numbers?

Reply:
To be quite honest, I have no clue.  They appear in ffmpeg.c without any documentation.